### PR TITLE
add monitoring module zabbix_host_triggers

### PR DESCRIPTION
**Pull request by stravassac**
##### SUMMARY
Add a module in monitoring zabbix to see trigger for a given Zabbix Host
Usefull to check the status of a host before run playbook on it

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
zabbix_host_triggers

### CODE REVIEW

**Review by D3DeFi — APPROVED** [review]
Looks good, thank you! :)

**Comment by D3DeFi** [comment]
shipit

**Comment by stravassac** [comment]
Thanks @D3DeFi

**Comment by stravassac** [comment]
Hi @D3DeFi do you know when this PR will be pushed in Ansible ?

**Comment by D3DeFi** [comment]
This PR affects version 2.9. As ansible follows [4 months release cycle](https://docs.ansible.com/ansible/latest/reference_appendices/release_and_maintenance.html) and 2.8 was released about a week ago, my guess is that this will ship some time in September (e.g. `pip install ansible==2.9`).

Until then, it will be accessible for users that are running ansible directly from source (`devel` branch) right after this PR gets merged. For that, we need a second shipit vote. Problem here is that I am currently only active (more like semi-active) maintainer for zabbix module namespace and PRs are usually waiting for 2nd vote from core maintainers. @resmo usually stops by and merges these things for which I am very grateful :) So maybe this will get merged sooner

**Comment by stravassac** [comment]
Thanks @D3DeFi for your reply

**Comment by D3DeFi** [comment]
cc @rubentsirunyan

File: lib/ansible/modules/monitoring/zabbix/zabbix_host_triggers.py
Comment (rubentsirunyan) [inline lib/ansible/modules/monitoring/zabbix/zabbix_host_triggers.py]: I believe this `else` block will never be executed, as `host_id_type` is defined with 3 choices and ansible will fail if something other than items in `choices` is passed to the module.
Even if this block is there for some reason:
1. I think the return message is not descriptive.
2. It would be nice to use Ansible's common `msg` instead of `result` as a return field for the message.

File: lib/ansible/modules/monitoring/zabbix/zabbix_host_triggers.py
Comment (rubentsirunyan) [inline lib/ansible/modules/monitoring/zabbix/zabbix_host_triggers.py]: I think `changed` is what you need here instead of `ok`. However as this module is a "_facts" module and it should never change anything on the remote machine, the value of `changed` should always be `False` (which i believe is the default behavior).

File: lib/ansible/modules/monitoring/zabbix/zabbix_host_triggers.py
Comment (rubentsirunyan) [inline lib/ansible/modules/monitoring/zabbix/zabbix_host_triggers.py]: Actually `output` field which contains `triggers_ok` and `triggers_problem` child elements is never returned. What you want here is having `triggers_ok` and `triggers_problem` fields at the root level with `type=json`.
Please see the documentation [here](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#return-block).

File: lib/ansible/modules/monitoring/zabbix/zabbix_host_triggers.py
Comment (rubentsirunyan) [inline lib/ansible/modules/monitoring/zabbix/zabbix_host_triggers.py]: Please specify the necessary fields for each return value as described in the [documentation](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#return-block), and maybe move all the fields under `triggers_ok` and `triggers_problem` to their corresponding `sample` field.

**Review by rubentsirunyan — CHANGES_REQUESTED** [review]
Thank you @stravassac for the contribution!
Please see my suggested changes in the comments.

Another thing I'd like to mention is that for me the name `zabbix_host_triggers` does not correctly describe what it does. It is more about [events](https://www.zabbix.com/documentation/4.2/manual/config/events) than [triggers](https://www.zabbix.com/documentation/4.2/manual/config/triggers) IMHO. @D3DeFi what are your thoughts on this?
And also a suffix like `_facts` would be nice to show that it does not supposed to do any changes.

**Comment by D3DeFi** [comment]
> 
> 
> Thank you @stravassac for the contribution!
> Please see my suggested changes in the comments.
> 
> Another thing I'd like to mention is that for me the name `zabbix_host_triggers` does not correctly describe what it does. It is more about [events](https://www.zabbix.com/documentation/4.2/manual/config/events) than [triggers](https://www.zabbix.com/documentation/4.2/manual/config/triggers) IMHO. @D3DeFi what are your thoughts on this?
> And also a suffix like `_facts` would be nice to show that it does not supposed to do any changes.

Yes, excellent idea @rubentsirunyan . A lot has changed since this was reviewed for the first time. Modules name as `_facts` should be used only if they are returning `ansible_facts` - see #56822 for reference. So maybe `zabbix_host_events_info` is a way to go?

There has been an additional things standardized:

- Zabbix modules should call `zbx.logout` before exit/fail - see #58525
- Failing with dependency error should be handled via helper function - see #58002

Would be cool if @stravassac can implement these changes as well, since we will still wait for some core dev to merge this.

**Comment by rubentsirunyan** [comment]
> Modules name as _facts should be used only if they are returning ansible_facts - see #56822 for reference.

Thanks @D3DeFi. Missed that one.

> So maybe zabbix_host_events_info is a way to go?

Seems a perfect name to me.

File: lib/ansible/modules/monitoring/zabbix/zabbix_host_triggers.py
Comment (stravassac) [reply lib/ansible/modules/monitoring/zabbix/zabbix_host_triggers.py]: @rubentsirunyan the "ok" keyword is like the description field "value" of trigger object (https://www.zabbix.com/documentation/3.0/manual/api/reference/trigger/object).

File: lib/ansible/modules/monitoring/zabbix/zabbix_host_triggers.py
Comment (stravassac) [reply lib/ansible/modules/monitoring/zabbix/zabbix_host_triggers.py]: @rubentsirunyan like this ?:
```
RETURN = '''
---
triggers_ok:
    description: Host Zabbix Triggers in OK state (https://www.zabbix.com/documentation/3.0/manual/api/reference/trigger/object)
    returned: On success
    type: complex
    contains:
            - comments: Additional description of the trigger
              description: Name of the trigger
              error: Error text if there have been any problems when updating the state of the trigger
              expression: Reduced trigger expression
              flags: Origin of the trigger
              lastchange: Time when the trigger last changed its state
              priority: Severity of the trigger
              state: State of the trigger
              status: Whether the trigger is enabled or disabled
              templateid: ID of the parent template trigger
              triggerid: ID of the trigger
              type: Whether the trigger can generate multiple problem events
              url: URL associated with the trigger
              value: Whether the trigger is in OK or problem state
triggers_problem: ........
```

Comment (D3DeFi) [reply lib/ansible/modules/monitoring/zabbix/zabbix_host_triggers.py]: yes, this looks better

File: lib/ansible/modules/monitoring/zabbix/zabbix_host_events_info.py
Comment (D3DeFi) [inline lib/ansible/modules/monitoring/zabbix/zabbix_host_events_info.py]:
```suggestion
module: zabbix_host_events_info
```

File: lib/ansible/modules/monitoring/zabbix/zabbix_host_events_info.py
Comment (D3DeFi) [inline lib/ansible/modules/monitoring/zabbix/zabbix_host_events_info.py]:
```suggestion
  zabbix_host_events_info:
```

File: lib/ansible/modules/monitoring/zabbix/zabbix_host_events_info.py
Comment (D3DeFi) [inline lib/ansible/modules/monitoring/zabbix/zabbix_host_events_info.py:157]: `type: str` missing under each of the options

File: lib/ansible/modules/monitoring/zabbix/zabbix_host_events_info.py
Comment (D3DeFi) [inline lib/ansible/modules/monitoring/zabbix/zabbix_host_events_info.py]:
```suggestion
version_added: "2.10"
```

File: lib/ansible/modules/monitoring/zabbix/zabbix_host_events_info.py
Comment (D3DeFi) [inline lib/ansible/modules/monitoring/zabbix/zabbix_host_events_info.py]:
```suggestion
                - alias: Account who acknowledge
```

File: lib/ansible/modules/monitoring/zabbix/zabbix_host_events_info.py
Comment (D3DeFi) [inline lib/ansible/modules/monitoring/zabbix/zabbix_host_events_info.py]: Ideally this should be covered in documentation of module so links are not necessary, but 
```suggestion
    description: Host Zabbix Triggers in problem state. See trigger and event objects in API documentation of your zabbix version for more
```

**Review by D3DeFi — CHANGES_REQUESTED** [review]
Please squash these commits under single commit that states correct module name.

**Comment by stravassac** [comment]
@D3DeFi thanks for your suggestion, I try to run "ansible-test sanity check" in local but I have some trouble.
Can you take  look

**Comment by D3DeFi** [comment]
> 
> 
> @D3DeFi thanks for your suggestion, I try to run "ansible-test sanity check" in local but I have some trouble.
> Can you take look

No need, It was already [tested for you with shippable](https://app.shippable.com/github/ansible/ansible/runs/143042/summary/console)

@gundalow can this be merged as well now please?

**Comment by stravassac** [comment]
Hello @D3DeFi, do you know why the merge stall ?

**Comment by D3DeFi** [comment]
None of the core devs stumbled upon this so far and I was only active maintainer at the time so it was missing one more vote to obtain shipit tag (bigger chance that someone from core will come and merge this as new modules are not allowed to be automerged).

@sky-joker can you please take a quick look at this module when you have a time. My review here is a bit outdated as of now.
@stravassac please [rebase ](https://docs.ansible.com/ansible/latest/dev_guide/developing_rebasing.html) this PR so stale_ci gets removed

**Review by D3DeFi — APPROVED** [review]
thanks!
@resmo can you merge please?

shipit

**Review by resmo — APPROVED** [review]
LGTM

**Comment by resmo** [comment]
rebase_merge

**Comment by resmo** [comment]
bot_status

**Comment by D3DeFi** [comment]
Thank you @resmo for merging! Thank you @stravassac for contrib! :)

**Pull request merged**
